src/Sarif/Readers/DeferredDictionary.cs

```diff
@@ -0,0 +1,514 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Sarif.Readers
+{
+    /// <summary>
+    ///  DeferredDictionary is an IDictionary&lt;string, T&gt; which uses JSON.NET
+    ///  to read the dictionary items as they are accessed. It pre-builds the mapping
+    ///  from string to JSON file position for fast retrieval.
+    ///  
+    ///  Enumerate KeyValuePairs for best performance:
+    ///  foreach(KeyValuePair&lt;T, U&gt;> item in dictionary)
+    ///  { ... }
+    /// </summary>
+    /// <remarks>
+    ///  DeferredDictionary doesn't record anything but the file position of the dictionary in the JSON initially.
+    ///  If you foreach over the KeyValuePairs, it constructs one reader and loads the items as you read them.
+    ///  If you ask for the Keys, Values, or Dictionary[key], it must build a map from each key to the file position for that value,
+    ///  and must seek in the file for each read.
+    ///  
+    ///  Items are expensive to iterate each time; they are not kept in memory. Copy the values to a List or array to keep them.
+    /// </remarks>
+    /// <typeparam name="T">Type of items in Dictionary</typeparam>
+    public class DeferredDictionary<T> : IDictionary<string, T>
+    {
+        private JsonSerializer _jsonSerializer;
+        private Func<Stream> _streamProvider;
+        private long _start;
+
+        private Stream _stream;
+        private Dictionary<int, long> _itemPositions;
+
+        public DeferredDictionary(JsonSerializer jsonSerializer, JsonPositionedTextReader reader, bool buildPositionsNow = false)
+        {
+            _jsonSerializer = jsonSerializer;
+            _streamProvider = reader.StreamProvider;
+            _start = reader.TokenPosition;
+
+            // We have the JsonTextReader, which must scan to after the collection to resume building the outer object
+            // We may as well make the map of element positions.
+            if (buildPositionsNow)
+            {
+                BuildPositions(reader, 0);
+            }
+            else
+            {
+                reader.Skip();
+            }
+        }
+
+        private void EnsurePositionsBuilt()
+        {
+            if (_itemPositions == null) BuildPositions();
```

Comment:
>BuildPositions [](start = 40, length = 14)

here's a nice nit for you, prefer adding braces even for single line if statements: 

if (_itemPostions == null) { BuildPositions(); }

Reply:
I fixed this class. I'm not sure if I'll fix all of them on this visit.